### PR TITLE
`modal shell` uses `container_exec`

### PR DESCRIPTION
**Pull request by aksh-at**
Noticed there was a lot of code duplication across `modal shell` and `container_exec`. The code was really hard to follow, with both calling `connect_to_terminal` eventually with two functions that had very specific call signatures (e.g. `connect_to_terminal` would call a function that was passed in with an `event` that it created). 

This PR makes it so that `modal shell` just does the equivalent of `container exec` in a separate process. 

Still not quite done:
- [x] Fix tests
- [x] Confirm error handling still works
- [x] Figure out what to run in the main container process — does `sleep infinity` work on all distros? It's possible that makes this approach unviable.

### CODE REVIEW

**Comment by erikbern** [comment]
I'm not familiar with this code but this definitely looks like a huge simplification so 👍

**Comment by aksh-at** [comment]
@prbot approve

**Comment by aksh-at** [comment]
@prbot approve

**Comment by erikbern** [comment]
👍
